File: csharp/saveImg/saveImg.cs
/*
saveImg.cs - Demonstrates how to use CSVizionSDK to get the image and save the image by OpenCV.
*/

using System;
using System.Runtime.InteropServices;
using OpenCvSharp;
using VizionSDKSharp;

class SaveImg
{
    static void Main()
    {
        // Discover camera devices
        StringVector devList = new StringVector();
        int camCount = CSVizionSDK.VxDiscoverCameraDevices(devList);
        if (camCount == 0)
        {
            Console.WriteLine("Cannot find any cameras!");
            return;
        }

        Console.WriteLine("Device List:");
        for (int i = 0; i < devList.Count; i++)
        {
            Console.WriteLine($"[{i}] {devList[i]}");
        }

        // Select first camera
        int devIdx = 0;
        Console.WriteLine($"Selected [{devIdx}] device");
        Console.WriteLine($"Device name: {devList[devIdx]}");

        var cam = CSVizionSDK.VxInitialCameraDevice(devIdx);

        // Open camera
        CSVizionSDK.VxOpen(cam);

        // Get available formats
        VxFormatVector fmtList = new VxFormatVector();
        CSVizionSDK.VxGetFormatList(cam, fmtList);

        int minWidth = 1920;
        int minHeight = 1080;
        VxFormat uyvyFmt = new VxFormat();

        foreach (var fmt in fmtList)
        {
            // Find minimum resolution in MJPG format
            if (fmt.format == VX_IMAGE_FORMAT.UYVY &&
                (fmt.width * fmt.height == minWidth * minHeight))
            {
                minWidth = fmt.width;
                minHeight = fmt.height;
                uyvyFmt = fmt;
            }
        }

        // Set capture format
        if (CSVizionSDK.VxSetFormat(cam, uyvyFmt) == 0)
            Console.WriteLine("Set Capture Format Success!");
        else
            Console.WriteLine("Set Capture Format Failed!");

        // Start streaming
        CSVizionSDK.VxStartStreaming(cam);

        // Get image
        int dataSize;
        byte[] buffer = new byte[minWidth * minHeight * 2];
        VX_CAPTURE_RESULT result = CSVizionSDK.VxGetImage(cam, buffer, out dataSize, 3000);
        if (result != VX_CAPTURE_RESULT.VX_SUCCESS)
        {
            Console.WriteLine("fail");
        }
        else
        {
            Console.WriteLine($"Get Image: {result} / {dataSize}");
        }

        // Create Mat from raw buffer
        using var matImg = new Mat(minHeight, minWidth, MatType.CV_8UC2);

        // Copy byte array into Mat
        Marshal.Copy(buffer, 0, matImg.Data, buffer.Length);

        // Convert from UYVY to BGR
        using var bgrImg = new Mat();
        Cv2.CvtColor(matImg, bgrImg, ColorConversionCodes.YUV2BGR_UYVY);

        // Save image
        Cv2.ImWrite("capture.png", bgrImg);

        // Close camera
        CSVizionSDK.VxStopStreaming(cam);
        CSVizionSDK.VxClose(cam);

        // Cleanup
        devList.Dispose();
        fmtList.Dispose();
    }
}

File: python/saveImg.py
"""
saveImg.py - Demonstrates how to use pyvizionsdk to get the image and save the image by OpenCV.
"""

import pyvizionsdk
from pyvizionsdk import VX_IMAGE_FORMAT

import cv2
import numpy as np

result, camera_list = pyvizionsdk.VxDiscoverCameraDevices()
if result < 1:
    print('No cameras discovered.')
    quit()

# print camera_list
print("Device List:\n")
for camera in camera_list:
    print(camera)

# initialize camera device
camera = pyvizionsdk.VxInitialCameraDevice(0)

# open camera
result = pyvizionsdk.VxOpen(camera)
print("Open camera return code:", result)

print("Choose a format:\n1. UYVY\n2. MJPG(Default)")
fmt_id = int(input("Please enter format (1 or 2): ") or "1")
select_format = VX_IMAGE_FORMAT.VX_IMAGE_FORMAT_UYVY if fmt_id == 1 else VX_IMAGE_FORMAT.VX_IMAGE_FORMAT_MJPG

# get format
result, format_list = pyvizionsdk.VxGetFormatList(camera)
cap_format = None
min_width = 640
min_height = 480
for format in format_list:
    if format.format == select_format:
        resolution = format.width * format.height
        if (format.width * format.height) == (min_width * min_height) :
            cap_format = format

# set format
result = pyvizionsdk.VxSetFormat(camera, cap_format)

# start streaming
result = pyvizionsdk.VxStartStreaming(camera)

# get image
result, image = pyvizionsdk.VxGetImage(camera, 1000, cap_format)

# retrieve the data to opencv and save with cv2.imwrite()
nparr = np.frombuffer(image, np.uint8)

if select_format == VX_IMAGE_FORMAT.VX_IMAGE_FORMAT_UYVY:
    uyvy = nparr.reshape((min_height, min_width, 2))
    bgr = cv2.cvtColor(uyvy, cv2.COLOR_YUV2BGR_UYVY)
    cv2.imwrite("capture.png", bgr)
elif select_format == VX_IMAGE_FORMAT.VX_IMAGE_FORMAT_MJPG:
    image = cv2.imdecode(nparr, cv2.IMREAD_COLOR)
    cv2.imwrite("capture.png", image)

# stop streaming
pyvizionsdk.VxStopStreaming(camera)
# close camera
pyvizionsdk.VxClose(camera)